File: main/src/cgeo/geocaching/sorting/CacheComparator.java
package cgeo.geocaching.sorting;

import cgeo.geocaching.models.Geocache;
import cgeo.geocaching.storage.SqlBuilder;

import androidx.annotation.NonNull;

import java.util.Comparator;
import java.util.List;

public interface CacheComparator extends Comparator<Geocache> {

    /**
     * returns {@code true} if this is a comparator that is used by the app automatically and was not selected by the
     * user.
     */
    boolean isAutoManaged();

    String getSortableSection(@NonNull Geocache cache);

    /** Sorts the given list of caches using this comparator. Respects implementations of {@link #beforeSort(List)} and{@link #afterSort(List)} */
    void sort(List<Geocache> list, boolean inverse);

    default void addSortToSql(final SqlBuilder sql, final boolean sortDesc) {
        //do nothing by default
    }



}

File: main-kotlin/src/main/kotlin/cgeo/geocaching/sorting/CacheComparator.kt
// Auto-converted from Java to Kotlin
// WARNING: This code requires manual review and likely has compilation errors
// Please review and fix:
// - Method signatures (parameter types, return types)
// - Field declarations without initialization
// - Static members (use companion object)
// - Try-catch-finally blocks
// - Generics syntax
// - Constructors
// - And more...

package cgeo.geocaching.sorting

import cgeo.geocaching.models.Geocache
import cgeo.geocaching.storage.SqlBuilder

import androidx.annotation.NonNull

import java.util.Comparator
import java.util.List

interface CacheComparator : Comparator()<Geocache> {

    String getSortableSection(Geocache cache)

    /** Sorts the given list of caches using this comparator. */
    Unit sort(List<Geocache> list)

    /**
     * Can optionally be overridden to perform preparation (e.g. caching of values) before sort of a list via {@link #sort(List)}
     */
    default Unit beforeSort(final List<Geocache> list) {
        //by default, do nothing
    }

    /**
     * Can optionally be overridden to perform cleanup (e.g. deleting cached values) before sort of a list via {@link #sort(List)}
     */
    default Unit afterSort(final List<Geocache> list) {
        //by default, do nothing
    }

    default Unit addSortToSql(final SqlBuilder sql, final Boolean sortDesc) {
        //do nothing by default
    }
}
